### PR TITLE
Set owner on pile creation

### LINKED ISSUE #114: Maximum size of a pile is only 2 when in a hand
If you set "stackOffset" to 0 in a hand, cards will obviously be piled upon each other like in a common holder. However, the maximum size of a pile seems to be 2 once it is in a hand. If you pull in more cards, they will still stack on each other visually, but a new pile will be created. So let's say, you have 6 cards on top of each other - it won't be a pile of 6, but three piles of 2 lying on each other.

This interaction with  "childrenPerOwner" is probably not surprising (not sure if it is intended, though?!). But in some games, it would come in handy if piling within a hand was possible. For example, in games where there are different currencies (let's say blue, red, green), and players are taking money from a face-down deck each turn; instead of having all individual money cards in your hand, it'd be more convenient if all blue (red, green) money cards were just stacked upon each other in three piles (= three hands)
a
.

### PR DESCRIPTION
**Pull request by robartsd**
This should fix some pile issues in hands. Should close #114

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3435/pr-test (or any other room on that server)

**Comment by robartsd** [comment]
Looks like I need to update the hash of game states for TestCafe, but I think it is actually working.

**Comment by bjalder26** [comment]
If you remove the pile from the holder by the handle, then the cards still retain their owner and are not visible by other players even after they exit the holder.

**Comment by robartsd** [comment]
> If you remove the pile from the holder by the handle, then the cards still retain their owner and are not visible by other players even after they exit the holder.

You're talking about behavior with this PR correct? That is a problem that needs attention.

**Comment by bjalder26** [comment]
> > If you remove the pile from the holder by the handle, then the cards still retain their owner and are not visible by other players even after they exit the holder.
> 
> You're talking about behavior with this PR correct? That is a problem that needs attention.

Yes, though technically it is also behavior on the live server as well.  In other words, it uncorrected from the main branch.

**Comment by robartsd** [comment]
Latest push breaks pile property changes.

**Comment by bjalder26** [comment]
Unfortunately, this recent commit crashes my client.  I'm certainly no expert, but if I were to make a guess, I would guess that this secondary issue could be solved by editing the dispenseCard function in holder.js so that when a card leaves it has its owner set to null.

**Comment by robartsd** [comment]
It was a typo in the name of the method. Looks like it is working now.

**Comment by bjalder26** [comment]
I refreshed and re-tested, and it appears you fixed that secondary issue as well.  I can't think of any other way to test this, so it seems to work perfectly to me.

**Comment by bjalder26** [comment]
In an attempt to discover errors in this PR I discovered that if you split a pile, then part of the pile can be removed from the holder while the ownership is retained (both the cards and the pile retain the same owner).  I don't know if that is a bug or a feature.  If it qualifies as a bug, I don't know if that should be covered in this PR.

**Comment by robartsd** [comment]
> In an attempt to discover errors in this PR I discovered that if you split a pile, then part of the pile can be removed from the holder while the ownership is retained (both the cards and the pile retain the same owner). I don't know if that is a bug or a feature. If it qualifies as a bug, I don't know if that should be covered in this PR.

We should test to see if other there are other effects of leaving a holder that get circumvented this way.

**Comment by bjalder26** [comment]
> We should test to see if other there are other effects of leaving a holder that get circumvented this way.
Good idea.  I did some testing, and this is what I found.

When pile carried by the handle
Cards in pile:
onEnter/onLeave applies to cards
enterRoutine/leaveRoutine does not collect them as a child

Piles:
enterRoutine/leaveRoutine collects them as a child
onEnter/onLeave is not applied
===============
Single card moved in and out:
enterRoutine/leaveRoutine collect as child
onEnter/onLeave apply to card
==================
Pile in holder split:
"Split out" portion
onEnter/onLeave applies to cards
leaveRoutine collects cards as children

NOTHING is applied to the new pile
(apparently, the new cards are moved out, then the pile forms)

Manually moved out portion
onEntry/onLeave applied to cards
enterRoutine/leaveRoutine collects pile as a child

**Comment by robartsd** [comment]
I tried to run TestCafe on a local server, but I got unexpected results (failure on my local machine was different from failure here).

```
$ npm run testcafe

> virtualtabletop@0.1.0 testcafe
> testcafe chrome tests/testcafe/tests.js

 Running tests in:
 - Chrome 91.0.4472.77 / Linux 0.0

 virtualtabletop.io
 ✖ Create game using edit mode

   1) A JavaScript error occurred on
      "http://localhost:8272/testcafe-testing".
      Repeat test actions in the browser and check the console for errors.
      To ignore client-side JavaScript errors, enable the "--skip-js-errors"
      CLI option, or set the "skipJsErrors" configuration file property to
      "true".
      If the website only throws this error when you test it with TestCafe,
      please create a new issue at:

   "https://github.com/DevExpress/testcafe/issues/new?template=bug-report.md".
      
      JavaScript error details:
      SyntaxError: Identifier 'a' has already been declared

      Browser: Chrome 91.0.4472.77 / Linux 0.0

         32 |    .click('p > .remove');
         33 |}
         34 |
         35 |async function setName(t, name) {
         36 |  await t
       > 37 |    .click('#playersButton')
         38 |    .typeText('.myPlayerEntry > .playerName', name ||
      'TestCafe', { replace: true })
         39 |    .click('#playersButton');
         40 |}
         41 |
         42 |function prepareClient() {
```
Each of the tests fails in this way on my server.

File: client/js/widgets/pile.js [hunk 1]
```diff
@@ -130,6 +130,14 @@ class Pile extends Widget {
       await this.parent.dispenseCard(child);
   }
 
+  async onPropertyChange(property, oldValue, newValue) {
+    if(property == 'owner') {
```
Comment (robartsd) [inline client/js/widgets/pile.js:134]: Perhaps we should consider fixing additional widget issues by passing through more property changes.

**Comment by robartsd** [comment]
Here's the diff for my TestCafe run:
```
 {
+  oklz: {
+    owner: "TestCafe"
+  }
   9eev_A_C: {
-    z: 3
+    z: 4
-    x: 90
+    x: 130
   }
   9eev_2_C: {
-    z: 2
+    z: 3
-    x: 50
+    x: 90
   }
   9eev_7_D: {
-    z: 6
+    z: 7
-    x: 210
+    x: 250
   }
   9eev_8_D: {
-    z: 8
+    z: 9
-    x: 290
+    x: 330
   }
   9eev_A_H: {
-    z: 7
+    z: 8
-    x: 250
+    x: 290
   }
   9eev_2_H: {
-    z: 10
+    z: 1
-    x: 370
+    x: 10
   }
   9eev_3_H: {
-    z: 4
+    z: 5
-    x: 130
+    x: 170
   }
   9eev_6_H: {
-    z: 9
+    z: 10
-    x: 330
+    x: 370
   }
   9eev_T_H: {
-    z: 1
+    z: 2
-    x: 10
+    x: 50
   }
   9eev_J_S: {
-    z: 5
+    z: 6
-    x: 170
+    x: 210
   }
 }
```
Looks like the cards in the hand are just in a different order: what was the last card is now the first.

**Comment by bjalder26** [comment]
>  it appears that the pile is not quite fully destroyed (parent property remains).


Could this be related to issue #477 and what I found in testing PR #521?

**Comment by robartsd** [comment]
Every time a pile that is owned is destroyed by removing the last widget, it gets left behind empty of all properties except "owner".

File: client/js/widgets/pile.js [hunk 2]
```diff
@@ -130,6 +130,14 @@ class Pile extends Widget {
       await this.parent.dispenseCard(child);
   }
 
+  async onPropertyChange(property, oldValue, newValue) {
+    if(this.children().length && property == 'owner') {
+      for(const c of this.children())
+        await c.set('owner', newValue);
+    }
+    super.onPropertyChange(property, oldValue, newValue);
```
Comment (ArnoldSmith86) [inline client/js/widgets/pile.js:138]: This probably needs an `await`? _Might_ actually be the reason for your fragments?

Comment (robartsd) [reply client/js/widgets/pile.js]: I'm not very well versed in the use of async and await. We'll se if this helps (it will be a few hours before I get back to my local server).

**Comment by bjalder26** [comment]
I'm getting lost in what piles should and shouldn't do so forgive me if some of this is not helpful.

One thing I noticed is that you can't SET to piles - it gets 'passed' to the children.  The only reason I can think of for SETting to piles would be testing, so this seems like desired behavior. 

When you place a pile inside of a holder, only the pile changes owner (which I suspect is also desired behavior).

When you stack a pile on top of a different pile, the piles do not change owner, one just gets deleted (which I think is desired behavior).

When I put a card on another card inside of a holder, one of the cards somehow appears to have its owner changed to the pile twice.  This does not happen when they are outside of a holder, and I don't know what is going on.
image

It is the top card (which I gave the id 'topcard') that has its owner changed twice.
image


Basically, the same thing happens when you set a pile on top of another pile inside of a holder.  One of the cards has 2 parent changes to the pile.
image

I can't think of anything else to test right now.  I set up a testing area in the room below.  The timer will SET a parentChangeRoutine on every widget every second (except piles, for which you'd have to manually set one).  The label on the right will show parent changes for every widget with the parentChangeRoutine.

**Comment by ArnoldSmith86** [comment]
`SET` should work for piles. `SELECT` replaces piles by their children by default but if you add `type: pile`, you should be able to select them (note that they still have to match `property`=`value`).

**Comment by bjalder26** [comment]
> `SET` should work for piles. `SELECT` replaces piles by their children by default but if you add `type: pile`, you should be able to select them (note that they still have to match `property`=`value`).

Yeah, that did work.  Now, as long as the timer is running in that room everything will have a parentChangeRoutine.  However, the only one odd thing I noticed, was when a card or cards are added to a pile (new or old) that is in a holder, the top card will change parents twice.

Also, I was wrong about the pile that gets destroyed not having a parent change.  It does, I just missed it before.

One card is added to a pile in a holder (changes parent twice):
image

Two cards are added to a pile in a holder (top card changes parents twice, the second card changes parent once, pile changes parent to the holder once):
image

Three cards added to a pile (same, but bottom card changes parents first):
image

For some reason the cards are 'processed' from bottom to top, except the card that is 2nd from the top is always LAST (then the pile).
image
image

**Comment by robartsd** [comment]
> When you place a pile inside of a holder, only the pile changes owner (which I suspect is also desired behavior).

Yes, the pile is directly in the holder, but the cards remain in the pile. If the pile is the holder's only child, the cards get listed as children of the holder internally, but the card's parent property still shows the pile.

You mention `owner` several times, but it seems you are talking about `parent`. This PR should not be changing behavior of piles with regard to `parent` property, only `owner` property should change with this PR.

**Comment by bjalder26** [comment]
> > When you place a pile inside of a holder, only the pile changes owner (which I suspect is also desired behavior).
> 
> Yes, the pile is directly in the holder, but the cards remain in the pile. If the pile is the holder's only child, the cards get listed as children of the holder internally, but the card's parent property still shows the pile.
> 
> You mention `owner` several times, but it seems you are talking about `parent`. This PR should not be changing behavior of piles with regard to `parent` property, only `owner` property should change with this PR.

Yeah sorry, parent, not owner.  I'm going to correct it above so it is less confusing.

**Comment by robartsd** [comment]
> > > When you place a pile inside of a holder, only the pile changes owner (which I suspect is also desired behavior).
> > 
> > 
> > Yes, the pile is directly in the holder, but the cards remain in the pile. If the pile is the holder's only child, the cards get listed as children of the holder internally, but the card's parent property still shows the pile.
> > You mention `owner` several times, but it seems you are talking about `parent`. This PR should not be changing behavior of piles with regard to `parent` property, only `owner` property should change with this PR.
> 
> Yeah sorry, parent, not owner. I'm going to correct it above so it is less confusing. I'm getting lost in all the things I'm doing.

For any parent issues you find, please report if this is a change in behavior from production or not. There should be no changes in parent behavior of piles in this PR vs production.

**Comment by robartsd** [comment]
> > > When you place a pile inside of a holder, only the pile changes owner (which I suspect is also desired behavior).
> > 
> > 
> > Yes, the pile is directly in the holder, but the cards remain in the pile. If the pile is the holder's only child, the cards get listed as children of the holder internally, but the card's parent property still shows the pile.
> > You mention `owner` several times, but it seems you are talking about `parent`. This PR should not be changing behavior of piles with regard to `parent` property, only `owner` property should change with this PR.
> 
> Yeah sorry, parent, not owner. I'm going to correct it above so it is less confusing. I'm getting lost in all the things I'm doing.

For any parent issues you find, please report if this is a change in behavior from production or not. There should be no changes in parent behavior of piles in this PR vs production.

**Comment by bjalder26** [comment]
You can still split a pile to get it out of a childPerOwner holder without changing its parent.  Then if you split that pile, both splits will retain the same owner.  However, if you start separating the pile into individual cards then all the cards will lose ownership except the last one.

If you manually change the ownership of a pile in the JSON, then it will not change the ownership of the cards in the pile.  HOWEVER, it will make the cards disappear from the view of non-owners.  I suspect this is as it should be.  A bunch of other things I tested was as expected.

**Comment by robartsd** [comment]
> You can still split a pile to get it out of a childPerOwner holder without changing its parent (why you would do this I don't know).

You mean without changing `owner`. This was reported early on and I currently feel like it is not worth fixing.

> Then if you split that pile, both splits will retain the same owner. However, if you start separating the pile into individual cards then all the cards will lose ownership except the last one.

I'll have to think about this behavior. I didn't think they would lose owner, but I really don't want all but one losing owner (should be consistent behavior for the last card in the pile).

> If you manually change the ownership of a pile in the JSON, then it will not change the ownership of the cards in the pile. HOWEVER, it will make the cards disappear from the view of non-owners. I suspect this is as it should be. A bunch of other things I tested was as expected.

This feels like the right behavior to me. Game play changes of ownership should be handled by the engine, but changes in the editor feel like they are designer responsibility here.

**Comment by bjalder26** [comment]
> You mean without changing `owner`. This was reported early on and I currently feel like it is not worth fixing.

Yes, owner... I don't know what's with me today.  I don't see why anyone would ever do that, but I figure I should report whatever I find.

> I'll have to think about this behavior. I didn't think they would lose owner, but I really don't want all but one losing owner (should be consistent behavior for the last card in the pile).

To be clear, this does not happen inside a holder, nor does it happen if you remove a holder from a parentPerOwner holder, because the pile loses its owner.  This would only happen if you spilt deck to get it out of an owned holder, or if a PILE had its owner set outside of a holder.

> This feels like the right behavior to me. Game play changes of ownership should be handled by the engine, but changes in the editor feel like they are designer responsibility here.

I agree.

**Review by bjalder26 — COMMENTED** [review]
This fixes issue #114.

"onLeave": {
    "owner": null
  },

**Review by bjalder26 — APPROVED** [review]
Approve based on testing.

**Comment by robartsd** [comment]
> This fixes issue #114.
> The issue of splitting a pile to remove it without changing owner can be fixed with a simple onLeave:
> "onLeave": {
> "owner": null
> },

Also production is not immune to this issue. On production you can currently form a pile of 2 cards in a hand with `alignChildren:false` then move that pile dropping in your hand to make a pile you own that you can then add cards to. If you split the pile, The cards leave your hand and form piles (but just like forming piles in your hand on production, the piles are not owned by you so you end up with multiple piles each with two cards). It's pretty clear to me that there will still be issues with piles after this PR is merged, but I'm not seeing any evidence of it generating new pile problems.
